t/lib/TestApp.pm

```diff
@@ -59,6 +59,33 @@ get '/read_session' => sub {
     "name='$name'"
 };
 
+# For testing whether we can *read* a session var within the after hook (but not
+# set, because it's too late by then
+get '/session/after_hook/read' => sub {
+    session after_hook => "value set in route";
+    return "Meh";
+};
+hook after => sub {
+    my $response = shift;
+    if (request->path eq '/session/after_hook/read') {
+        $response->content("Read " . session('after_hook'));
+    }
+};
+
+# But we can't *set* a session var in the after hook, as the headers have been
+# built
+get '/session/after_hook/write' => sub {
+    session after_hook => "value set in route";
+    return "Meh";
+};
+hook after => sub {
+    my $response = shift;
+    if (request->path eq '/session/after_hook/write') {
+        session after_hook => "value changed in hook";
+        $response->content("Read " . session('after_hook'));
+    }
+};
```

Comment:
If you want to have two routes, one of which tries to alter the session in an after hook and one of which doesn't, why not define (a) standard code which matches two routes, (b) an after hook that modifies the response for both routes, and then (c) a further after hook that fires only for one of the routes?

Reply:
More importantly: if you revert the change to Dancer::Session::Abstract, do the tests still pass? If they don't, I'm very confused, because it looks like all you're doing is messing with an in-memory structure and then immediately returning part of it, which should obviously succeed. Do you perhaps need another route that just returns the contents of the session?

Reply:
The issue was that reading from the session would as a side effect generate a new cookie and push it into the response's cookies (overwriting any that we'd already put there).  That's a little inefficient, but more importantly means that if you read from a session in an after hook (e.g. as my code at $work was doing to pull a value from the session for logging purposes) the attempt to push the cookie again would explode - even though there was no need to try to update it.

However, looking at this again, I should add an additional test that makes sure that the changes made to the session did persist in the session, rather than not explode but not be flushed either (my concern is that the session stuff may also use an `after` hook to flush the session contents to the appropriate storage, and if so, that could fire before our `after` hook, in which case any changes we make would be silently discarded, which would be Not Good.